### PR TITLE
LPS-136518 Fix URLs in Guest site

### PR DESCRIPTION
**Pull request by lfbesada**
https://issues.liferay.com/browse/LPS-136518

This bug was first resolved but later was reverted because it was causing another bug. That is the reason why I am cherry-picking the original commits first and then solving the issues and adding tests.

cc: @jkappler @balazssk

### CODE REVIEW

**Comment by lfbesada** [comment]
ci:test:sf

**Comment by lfbesada** [comment]
ci:test:relevant

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 1]
```diff
@@ -60,32 +67,45 @@
 
 	@BeforeClass
 	public static void setUpClass() throws Exception {
+		_originalVirtualHostDefaultSiteName =
+			ReflectionTestUtil.getAndSetFieldValue(
+				PropsValues.class, "VIRTUAL_HOSTS_DEFAULT_SITE_NAME", "Guest");
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:72]: @lfbesada GroupConstants.GUEST

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 2]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		Layout nondefaultSiteLayout = LayoutTestUtil.addLayout(group);
+
+		_testGroupFriendlyURL(
+			_company.getVirtualHostname(),
+			PropsValues.LAYOUT_FRIENDLY_URL_PUBLIC_SERVLET_MAPPING +
+				group.getFriendlyURL(),
+			group, nondefaultSiteLayout);
+
+		String hostName =
+			RandomTestUtil.randomString(6) + "." +
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:171]: StringPool.PERIOD or CharPool.PERIOD

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 3]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		throws Exception {
+
+		try (SafeCloseable safeCloseable =
+				PropsValuesTestUtil.swapWithSafeCloseable(
+					"VIRTUAL_HOSTS_DEFAULT_SITE_NAME", StringPool.BLANK)) {
+
+			Layout defaultSiteLayout = _layoutLocalService.fetchDefaultLayout(
+				_group.getGroupId(), false);
+
+			_updateLayoutSetVirtualHostname(defaultSiteLayout, "");
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:191]: StringPool.BLANK

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 4]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+			Layout nondefaultSiteLayout = LayoutTestUtil.addLayout(group);
+
+			_testGroupFriendlyURL(
+				_company.getVirtualHostname(),
+				PropsValues.LAYOUT_FRIENDLY_URL_PUBLIC_SERVLET_MAPPING +
+					group.getFriendlyURL(),
+				group, nondefaultSiteLayout);
+
+			String hostName =
+				RandomTestUtil.randomString(6) + "." +
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:226]: StringPool.PERIOD or CharPool.PERIOD

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java [hunk 5]
```diff
@@ -8420,12 +8420,12 @@ private String _getGroupFriendlyURL(
 			}
 		}
 
-		if (useGroupVirtualHostname) {
-			TreeMap<String, String> virtualHostnames = getVirtualHostnames(
-				layoutSet);
+		String portalDomain = themeDisplay.getPortalDomain();
 
-			String portalDomain = themeDisplay.getPortalDomain();
+		TreeMap<String, String> virtualHostnames = getVirtualHostnames(
```
Comment (ealonso) [inline portal-impl/src/com/liferay/portal/util/PortalImpl.java:8425]: Map<String, String>

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 6]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		throws Exception {
 
-		LayoutSet layoutSet = layout.getLayoutSet();
+		Layout defaultSiteLayout = _layoutLocalService.fetchDefaultLayout(
+			_group.getGroupId(), false);
 
-		Map<String, String> virtualHostnames = TreeMapBuilder.put(
-			layoutHostname, StringPool.BLANK
-		).build();
+		_updateLayoutSetVirtualHostname(defaultSiteLayout, "");
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:140]: StringPool.BLANK

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java (hunk 1 above)
Comment (ealonso) [reply modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:72]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 7]
```diff
@@ -60,32 +67,45 @@
...
+		}
 
-		_updateLayoutSetVirtualHostname(_publicLayout, _PUBLIC_LAYOUT_HOSTNAME);
+		LayoutTestUtil.addLayout(_group, true);
 	}
 
 	@AfterClass
 	public static void tearDownClass() throws PortalException {
+		ReflectionTestUtil.setFieldValue(
+			PropsValues.class, "VIRTUAL_HOSTS_DEFAULT_SITE_NAME",
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:97]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 8]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		User user = UserTestUtil.getAdminUser(_company.getCompanyId());
+
+		Group group = GroupTestUtil.addGroup(
+			_company.getCompanyId(), user.getUserId(), 0);
+
+		Layout nondefaultSiteLayout = LayoutTestUtil.addLayout(group);
+
+		_testGroupFriendlyURL(
+			_company.getVirtualHostname(),
+			PropsValues.LAYOUT_FRIENDLY_URL_PUBLIC_SERVLET_MAPPING +
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:166]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 9]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		_testGroupFriendlyURL(
+			_PUBLIC_LAYOUT_HOSTNAME, StringPool.BLANK, _group,
+			defaultSiteLayout);
+
+		User user = UserTestUtil.getAdminUser(_company.getCompanyId());
+
+		Group group = GroupTestUtil.addGroup(
+			_company.getCompanyId(), user.getUserId(), 0);
+
+		Layout nondefaultSiteLayout = LayoutTestUtil.addLayout(group);
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:162]: nonDefaultSiteLayout

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 10]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+		_testGroupFriendlyURL(
+			hostName, StringPool.BLANK, group, nondefaultSiteLayout);
+	}
+
+	@Test
+	public void testGetGroupFriendlyURLFromPublicLayoutNoVirtualHostDefaultSite()
+		throws Exception {
+
+		try (SafeCloseable safeCloseable =
+				PropsValuesTestUtil.swapWithSafeCloseable(
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:185]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 11]
```diff
@@ -94,30 +114,123 @@ public void testGetGroupFriendlyURLFromPrivateLayout() throws Exception {
...
+			_testGroupFriendlyURL(
+				_PUBLIC_LAYOUT_HOSTNAME, StringPool.BLANK, _group,
+				defaultSiteLayout);
+
+			User user = UserTestUtil.getAdminUser(_company.getCompanyId());
+
+			Group group = GroupTestUtil.addGroup(
+				_company.getCompanyId(), user.getUserId(), 0);
+
+			Layout nondefaultSiteLayout = LayoutTestUtil.addLayout(group);
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:217]: nonDefaultSiteLayout

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 12]
```diff
@@ -141,6 +254,25 @@ private void _testGroupFriendlyURL(
...
+
+		LayoutSet layoutSet = layout.getLayoutSet();
+
+		Map<String, String> virtualHostnames = TreeMapBuilder.put(
+			layoutHostname, StringPool.BLANK
+		).build();
+
+		_virtualHostLocalService.updateVirtualHosts(
+			_company.getCompanyId(), layoutSet.getLayoutSetId(),
+			(TreeMap<String, String>)virtualHostnames);
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:268]: Map<String, String>

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 13]
```diff
@@ -141,6 +254,25 @@ private void _testGroupFriendlyURL(
...
+
+		layoutSet.setVirtualHostnames(
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:270]: Map<String, String>

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java [hunk 14]
```diff
@@ -141,6 +254,25 @@ private void _testGroupFriendlyURL(
 				layout.getLayoutSet(), themeDisplay, false, true));
 	}
 
+	private void _updateLayoutSetVirtualHostname(
+		Layout layout, String layoutHostname) {
+
+		LayoutSet layoutSet = layout.getLayoutSet();
+
+		Map<String, String> virtualHostnames = TreeMapBuilder.put(
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:262]: Why TreeMap and not HashMap?

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplLayoutFriendlyURLTest.java [hunk 15]
```diff
@@ -79,12 +83,32 @@ public void testCompanyDefaultSiteVirtualHost() throws Exception {
 	public void testCompanyDefaultSiteVirtualHostWithLayoutSetVirtualHost()
 		throws Exception {
 
-		_setLayoutSetVirtualHost();
+		try (SafeCloseable safeCloseable =
+				PropsValuesTestUtil.swapWithSafeCloseable(
+					"VIRTUAL_HOSTS_DEFAULT_SITE_NAME", GroupConstants.GUEST)) {
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplLayoutFriendlyURLTest.java:88]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplLayoutFriendlyURLTest.java [hunk 16]
```diff
@@ -79,12 +83,32 @@ public void testCompanyDefaultSiteVirtualHost() throws Exception {
...
+				_company.getVirtualHostname(), _layout.getFriendlyURL());
+		}
+	}
+
+	@Test
+	public void testCompanyNoDefaultSiteVirtualHostWithLayoutSetVirtualHost()
+		throws Exception {
+
+		try (SafeCloseable safeCloseable =
+				PropsValuesTestUtil.swapWithSafeCloseable(
```
Comment (ealonso) [inline modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplLayoutFriendlyURLTest.java:102]: PropsKeys.VIRTUAL_HOSTS_DEFAULT_SITE_NAME

**Review by ealonso — CHANGES_REQUESTED** [review]

File: portal-impl/src/com/liferay/portal/util/PortalImpl.java (hunk 5 above)
Comment (lfbesada) [reply portal-impl/src/com/liferay/portal/util/PortalImpl.java:8425]: See: https://github.com/brianchandotcom/liferay-portal/pull/113569/commits/c8ad37559d81d325f36326e94783a9fc508a36bb

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java (hunk 13 above)
Comment (lfbesada) [reply modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:270]: See: https://github.com/brianchandotcom/liferay-portal/pull/113569/commits/c8ad37559d81d325f36326e94783a9fc508a36bb

File: modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java (hunk 12 above)
Comment (lfbesada) [reply modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplGroupFriendlyURLTest.java:268]: See: https://github.com/brianchandotcom/liferay-portal/pull/113569/commits/c8ad37559d81d325f36326e94783a9fc508a36bb

**Comment by lfbesada** [comment]
See: https://github.com/liferay-echo/liferay-portal/pull/7323

**Pull request closed without merging**